File: samples/android/app/src/main/java/com/amazon/sampleapp/impl/NavigationHandler.java
/*
 * Copyright 2017-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazon.sampleapp.impl;

import android.content.Context;
import android.support.v4.content.ContextCompat;

import com.amazon.aace.logger.Logger;
import com.amazon.aace.navigation.Navigation;
import com.amazon.sampleapp.R;
import com.amazon.sampleapp.view.LogEntry;

import org.json.JSONObject;

import java.util.Observable;
import java.util.Observer;

public class NavigationHandler extends Navigation
{
    private LoggerHandler mLogger = null;
    private int mLogColor;
    static NavigationObservable sNavigationObservable = null;

    public NavigationHandler( Context context, LoggerHandler logger )
    {
        mLogger = logger;
        mLogColor = ContextCompat.getColor( context, R.color.logTextTemplate ) & 0x00FFFFFF;
    }

    @Override
    public boolean setDestination( String payload )
    {
        try
        {
            // Log payload
            JSONObject template = new JSONObject( payload );
            mLogger.post( template.toString( 4 ), Logger.Level.INFO, mLogColor );

            // Log display card
            JSONObject json = new JSONObject();
            json.put( "template", template );
            json.put( "source", LoggerHandler.Source.CLI.toString() );
            json.put( "level", Logger.Level.INFO.toString() );
            sNavigationObservable.log( "setDestinationTemplate", json );
            return true;
        }
        catch( Throwable ex )
        {
            ex.printStackTrace();
            return false;
        }
    }

    //
    // NavigationObservable
    //
    static public void addNavigationObserver( Observer observer )
    {
        if( sNavigationObservable == null ) {
            sNavigationObservable = new NavigationHandler.NavigationObservable();
        }

        sNavigationObservable.addObserver( observer );
    }

    static private class NavigationObservable extends Observable
    {
        public void log( String message ) {
            setChanged();
            notifyObservers( message );
        }

        public void log( String type, JSONObject obj )
        {
            setChanged();
            notifyObservers( new LogEntry( type, obj ) );
        }
    }
}

File: platforms/agl/alexa-voiceagent-service/src/plugins/aac-services/aasb/aac-platform/alexa/src/NavigationHandler.cpp
/*
 * Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#include "NavigationHandler.h"

#include <aasb/Consts.h>

#include <rapidjson/document.h>
#include <rapidjson/prettywriter.h>
#include <rapidjson/stringbuffer.h>

/**
 * Specifies the severity level of a log message
 * @sa @c aace::logger::LoggerEngineInterface::Level
 */
using Level = aace::logger::LoggerEngineInterface::Level;

// Namespaces.
using namespace aasb::bridge;

// Logging tag for this file.
const std::string TAG = "aasb::navigation::NavigationHandler";

namespace aasb {
namespace navigation {

std::shared_ptr<NavigationHandler>NavigationHandler::create(
    std::shared_ptr<aasb::core::logger::LoggerHandler> logger,
    std::weak_ptr<aasb::bridge::ResponseDispatcher> responseDispatcher) {
    return std::shared_ptr<NavigationHandler>(new NavigationHandler(logger, responseDispatcher));
}

NavigationHandler::NavigationHandler(
    std::shared_ptr<aasb::core::logger::LoggerHandler> logger,
    std::weak_ptr<aasb::bridge::ResponseDispatcher> responseDispatcher) :
        m_navigationState(""),
        m_logger(logger),
        m_responseDispatcher(responseDispatcher) {
    m_navigationState = createNavigationState("NOT_NAVIGATING");
}

bool NavigationHandler::setDestination( const std::string& payload ) {
    m_logger->log(Level::VERBOSE, TAG, "setDestination");

    auto responseDispatcher = m_responseDispatcher.lock();
    if (!responseDispatcher) {
        m_logger->log(Level::WARN, TAG, "setDestination: Directive dispatcher is out of scope");
        return false;
    }

    responseDispatcher->sendDirective(TOPIC_NAVIGATION, ACTION_NAVIGATION_SET_DESTINATION, payload);
    m_navigationState = createNavigationState("NAVIGATING");

    return true;
}

bool NavigationHandler::cancelNavigation() {
    m_logger->log(Level::VERBOSE, TAG, "cancelNavigation");

    auto responseDispatcher = m_responseDispatcher.lock();
    if (!responseDispatcher) {
        m_logger->log(Level::WARN, TAG, "cancelNavigation: Directive dispatcher is out of scope");
        return false;
    }

    responseDispatcher->sendDirective(TOPIC_NAVIGATION, ACTION_NAVIGATION_CANCEL, "");
    m_navigationState = createNavigationState("NOT_NAVIGATING");

    return true;
}

std::string NavigationHandler::getNavigationState() {
    m_logger->log(Level::VERBOSE,TAG, "getNavigationState: " + m_navigationState);

    return m_navigationState;
}

std::string NavigationHandler::createNavigationState(std::string state) {
    rapidjson::Document document;
    document.SetObject();

    // Add state
    document.AddMember(
        "state",
        rapidjson::Value().SetString(state.c_str(), state.length()),
        document.GetAllocator());

    // Add waypoints
    rapidjson::Value wayPoints(rapidjson::kArrayType);
    document.AddMember(
        "waypoints",
        wayPoints,
        document.GetAllocator());

    // Add shapes
    rapidjson::Value shapes(rapidjson::kArrayType);
    document.AddMember(
        "shapes",
        shapes,
        document.GetAllocator());

    // create event string
    rapidjson::StringBuffer buffer;
    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buffer);
    document.Accept(writer);

    return buffer.GetString();
}

}  // phoneCallController
}  // aasb